### PR TITLE
SAK-50440 Rubrics remove sharing of public rubrics from other sites

### PR DESCRIPTION
**Pull request by Aniii10**
In the rubrics section, the public rubrics that are displayed must be able to be unpublished or deleted from any site other than the original.  This is important for institutions with hundreds of shared rubrics or when the original instructor no longer belongs to the institution.

The development has been sponsored by Hotelschool.

### CODE REVIEW

**Comment by bgarciaentornos** [comment]
Leaving this as a draft, as Christina requested that these actions should be shown only to superusers. We're working on that but we'd like to have this included in 25 so creating the PR before the freeze date.

File: webcomponents/tool/src/main/frontend/packages/sakai-rubrics/src/SakaiRubricsSharedList.js [hunk 1]
```diff
@@ -34,18 +45,43 @@ export class SakaiRubricsSharedList extends SakaiRubricsList {
...
   }
 
   refresh() {
     this.getSharedRubrics();
   }
 
-  getSharedRubrics() {
+  refreshPage() {
+    window.location.reload();
+  }
```
Comment (bgarciaentornos) [inline webcomponents/tool/src/main/frontend/packages/sakai-rubrics/src/SakaiRubricsSharedList.js:82]: This method is no longer used and it could be removed

**Review by bgarciaentornos — CHANGES_REQUESTED** [review]

File: rubrics/api/src/main/bundle/rubrics_eu.properties [hunk 2]
```diff
@@ -19,6 +19,7 @@ save=Gorde
 done=Egina
 total=Guztira
 confirm_remove=Ziur zaude kendu egin nahi duzula?
+confirm_remove_public=Ziur "{}" publikoki partekatutako errubriketatik eta webgunearen jatorritik kendu nahi duzula?
```
Comment (bgarciaentornos) [inline rubrics/api/src/main/bundle/rubrics_eu.properties:22]: @pablosanroman can you check this, please?

File: webcomponents/tool/src/main/frontend/packages/sakai-rubrics/src/SakaiRubricReadonly.js [hunk 3]
```diff
@@ -8,14 +8,36 @@ export class SakaiRubricReadonly extends SakaiRubric {
...
   constructor() {
 
     super();
-
+    this.isSuperUser = false;
     this.enablePdfExport = false;
+    this.fetchUserRoles();
+  }
+
+  async fetchUserRoles() {
```
Comment (bgarciaentornos) [inline webcomponents/tool/src/main/frontend/packages/sakai-rubrics/src/SakaiRubricReadonly.js:23]: Please rename this to checkSuperUser (we're not actually getting other roles)

File: webapi/src/main/java/org/sakaiproject/webapi/controllers/UserController.java [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+
+@RestController
+@Slf4j
+public class UserController extends AbstractSakaiApiController{
+
+    @Autowired
+    private SecurityService securityService;
+
+    @GetMapping("/user/roles")
+    public ResponseEntity<Map<String, Boolean>> getUserRoles() {
```
Comment (bgarciaentornos) [inline webapi/src/main/java/org/sakaiproject/webapi/controllers/UserController.java:22]: Please rename this to checkSuperUser (we're not actually getting other roles)

File: webcomponents/tool/src/main/frontend/packages/sakai-rubrics/src/SakaiRubricReadonly.js [hunk 5]
```diff
@@ -8,14 +8,36 @@ export class SakaiRubricReadonly extends SakaiRubric {
...
+      });
+
+      if (!response.ok) {
+        throw new Error(`HTTP error! status: ${response.status}`);
+      }
+
+      const roles = await response.json();
+      this.isSuperUser = roles.isSuperUser;
+
+      console.log("Is SuperUser:", this.isSuperUser);
```
Comment (bgarciaentornos) [inline webcomponents/tool/src/main/frontend/packages/sakai-rubrics/src/SakaiRubricReadonly.js:37]:
```suggestion
      console.debug("Is SuperUser:", this.isSuperUser);
```

File: rubrics/api/src/main/bundle/rubrics_eu.properties (hunk 2 above)
Comment (PabloSanRoman) [reply rubrics/api/src/main/bundle/rubrics_eu.properties:22]: Hi @bgarciaentornos!
Best with a little change: 
+ confirm_remove_public= Ziur kendu nahi duzula errubrika hau, "{}", publikoki partekatutako errubriketatik eta gunetik bertatik?

**Review by bgarciaentornos — APPROVED** [review]
@adrianfish  Can you take a look, please?

File: rubrics/api/src/main/bundle/rubrics.properties [hunk 6]
```diff
@@ -21,6 +21,7 @@ save=Save
 done=Done
 total=Total
 confirm_remove=Are you sure you want to remove "{}" ?
+confirm_remove_public=Are you sure you want to remove "{}" both from publicly shared rubrics and from its site origin?
```
Comment (adrianfish) [inline rubrics/api/src/main/bundle/rubrics.properties:24]: I'm not sure about the word "public". This implies that these rubrics are visible to people who are not Sakai users, the public.

Comment (adrianfish) [inline rubrics/api/src/main/bundle/rubrics.properties:24]:
```suggestion
confirm_remove_shared=Are you sure you want to remove "{}" both from shared rubrics and from its source site?
```

File: rubrics/api/src/main/bundle/rubrics_ca.properties [hunk 7]
```diff
@@ -18,6 +18,7 @@ save=Desa
 done=Fet
 total=Total
 confirm_remove=Esteu segur de voler-ho eliminar
+confirm_remove_public=Esteu segur que voleu eliminar "{}" tant de les r\u00fabriques compartides p\u00fablicament com del seu origen del lloc?
```
Comment (adrianfish) [inline rubrics/api/src/main/bundle/rubrics_ca.properties:21]:
```suggestion
confirm_remove_shared=Esteu segur que voleu eliminar "{}" tant de les r\u00fabriques compartides p\u00fablicament com del seu origen del lloc?
```

File: rubrics/api/src/main/bundle/rubrics_es.properties [hunk 8]
```diff
@@ -20,6 +20,7 @@ save=Guardar
 done=Aceptar
 total=Total
 confirm_remove=\u00bfEst\u00e1s seguro de que deseas borrar "{}"?
+confirm_remove_public=\u00bfEst\u00e1s seguro de que deseas borrar "{}", tanto de las r\u00fabricas compartidas p\u00fablicamente como del sitio de origen?
```
Comment (adrianfish) [inline rubrics/api/src/main/bundle/rubrics_es.properties:23]:
```suggestion
confirm_remove_shared=\u00bfEst\u00e1s seguro de que deseas borrar "{}", tanto de las r\u00fabricas compartidas p\u00fablicamente como del sitio de origen?
```

File: rubrics/api/src/main/bundle/rubrics_eu.properties (hunk 2 above)
Comment (adrianfish) [inline rubrics/api/src/main/bundle/rubrics_eu.properties:22]:
```suggestion
confirm_remove_shared=Ziur "{}" publikoki partekatutako errubriketatik eta webgunearen jatorritik kendu nahi duzula?
```

File: webapi/src/main/java/org/sakaiproject/webapi/controllers/UserController.java [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+
+    @GetMapping("/user/roles")
+    public ResponseEntity<Map<String, Boolean>> checkSuperUser() {
+
+        Map<String, Boolean> roles = new HashMap<>();
+        roles.put("isSuperUser", securityService.isSuperUser());
+
+        return ResponseEntity.ok()
+          .contentType(MediaType.APPLICATION_JSON)
+          .body(roles);
```
Comment (adrianfish) [inline webapi/src/main/java/org/sakaiproject/webapi/controllers/UserController.java:29]:
```suggestion
        return ResponseEntity.ok()
          .contentType(MediaType.APPLICATION_JSON)
          .body(Map.of("isSuperUser", securityService.isSuperUser()));
```

File: webcomponents/tool/src/main/frontend/packages/sakai-rubrics/src/SakaiRubricReadonly.js [hunk 10]
```diff
@@ -8,14 +8,36 @@ export class SakaiRubricReadonly extends SakaiRubric {
...
     enablePdfExport: { attribute: "enable-pdf-export", type: Boolean },
   };
 
   constructor() {
 
     super();
-
+    this.isSuperUser = false;
     this.enablePdfExport = false;
+    this.checkSuperUser();
```
Comment (adrianfish) [inline webcomponents/tool/src/main/frontend/packages/sakai-rubrics/src/SakaiRubricReadonly.js:20]: This is a really bad idea. If we have 100 shared rubrics on a page, that's 100 fetch requests. Instead, you should think in terms of setting is-super-user on the actual tag as an attribute and supplying that once. Look in the RubricsController and look at how enablePdfExport is supplied from the controller. You will need to set that attribute on the list and shared list components so they can then set it on the readonly rubric tags. But, that is a lot better than firing off 100s of fetch requests.

**Review by adrianfish — CHANGES_REQUESTED** [review]

**Review by bgarciaentornos — CHANGES_REQUESTED** [review]
Please, apply requested changes

File: webcomponents/tool/src/main/frontend/packages/sakai-rubrics/src/SakaiRubricsManager.js [hunk 11]
```diff
@@ -19,6 +20,15 @@ export class SakaiRubricsManager extends RubricsElement {
...
     this.sharedRubricsExpanded = "false";
     this.enablePdfExport = false;
+    this.isSuperUser = false;
+  }
+
+  attributeChangedCallback(name, oldValue, newValue) {
+    super.attributeChangedCallback(name, oldValue, newValue);
+
+    if (name === "is-super-user") {
+      this.isSuperUser = newValue === "true";
```
Comment (adrianfish) [inline webcomponents/tool/src/main/frontend/packages/sakai-rubrics/src/SakaiRubricsManager.js:30]: Why did you need to do this? This seems a bit strange. You should just be able to let lit's setter sort this out from from the attribute.

**Review by adrianfish — CHANGES_REQUESTED** [review]

File: webcomponents/tool/src/main/frontend/packages/sakai-rubrics/src/SakaiRubricsManager.js (hunk 11 above)
Comment (Aniii10) [reply webcomponents/tool/src/main/frontend/packages/sakai-rubrics/src/SakaiRubricsManager.js:30]: Yes, you're right. The reason I did it this way was to handle converting the is-super-user attribute (which is always a string in HTML) into a boolean (true or false) value for the property. But I realize that Lit can handle this automatically if we use its built-in property reflection. So I'm going to modify it. Thanks a lot for pointing that out.

Comment (Aniii10) [reply webcomponents/tool/src/main/frontend/packages/sakai-rubrics/src/SakaiRubricsManager.js:30]: Hi Adrian, I’ve been reviewing the implementation, and after considering the collaboration and recommendations from LIT contributors, I believe it’s best to keep the attributeChangedCallback method as it is.

The main reason for this is that in LIT, boolean attributes are interpreted solely based on their presence in the DOM. This means that even if the is-super-user attribute is passed as false, the component will interpret it as true simply because the attribute exists. To avoid confusion and ensure the component’s logic is accurate, it’s necessary to convert it to a real boolean value.

By using the attributeChangedCallback method, we can ensure that isSuperUser is stored correctly as a real boolean with this.isSuperUser = newValue === "true";.

Comment (adrianfish) [reply webcomponents/tool/src/main/frontend/packages/sakai-rubrics/src/SakaiRubricsManager.js:30]: It's a shame we can't make Thymeleaf conditionally add that boolean attribute. Boolean attributes are not supposed to be set to a value because that just doesn't make sense. However, I see the issue. Thymeleaf just won't let you conditionally add an attribute unless it is one of the standard dialect th:checked type attributes.

Comment (adrianfish) [reply webcomponents/tool/src/main/frontend/packages/sakai-rubrics/src/SakaiRubricsManager.js:30]: However ...

https://stackoverflow.com/questions/64803308/thymeleaf-add-attribute-without-value

Comment (Aniii10) [reply webcomponents/tool/src/main/frontend/packages/sakai-rubrics/src/SakaiRubricsManager.js]: That way works perfectly, and there's no need to use the attributeChangedCallback method. Thank you for sharing the StackOverflow link!

I encountered the same issue with enablePdfExport, and I resolved it in the same way as with isSuperUser. I agree that the problem lies with Thymeleaf not allowing custom boolean attributes, so putting the hacky code in the Thymeleaf template makes sense. Thanks again!

**Comment by bgarciaentornos** [comment]
@adrianfish Could you please take another look? Thanks!

**Review by bgarciaentornos — APPROVED** [review]

File: webcomponents/tool/src/main/frontend/packages/sakai-rubrics/src/SakaiRubricsSharedList.js [hunk 12]
```diff
@@ -45,7 +80,6 @@ export class SakaiRubricsSharedList extends SakaiRubricsList {
   }
 
   getSharedRubrics() {
-
```
Comment (adrianfish) [inline webcomponents/tool/src/main/frontend/packages/sakai-rubrics/src/SakaiRubricsSharedList.js:48]: Please leave these blank lines in. I know it seems a bit pedantic, but I think it makes js easier to read to leave a blank like on a multiline function definition - I do that everywhere in the webcomponents stuff.

File: webcomponents/tool/src/main/frontend/packages/sakai-rubrics/src/SakaiRubricsSharedList.js [hunk 13]
```diff
@@ -59,12 +93,60 @@ export class SakaiRubricsSharedList extends SakaiRubricsList {
     .catch (error => console.error(error));
   }
 
+  showDeleteModal(e) {
+    e.stopPropagation();
```
Comment (adrianfish) [inline webcomponents/tool/src/main/frontend/packages/sakai-rubrics/src/SakaiRubricsSharedList.js:97]:
```suggestion

    e.stopPropagation();
```

File: webcomponents/tool/src/main/frontend/packages/sakai-rubrics/src/SakaiRubricsSharedList.js [hunk 14]
```diff
@@ -59,12 +93,60 @@ export class SakaiRubricsSharedList extends SakaiRubricsList {
...
+    modal.show();
+  }
   copyToSite(e) {
 
     SakaiRubricsHelpers.get(`/api/sites/${this.siteId}/rubrics/${e.detail}/copyToSite`, {})
       .then(() => this.dispatchEvent(new CustomEvent("copy-share-site")));
   }
 
+  confirmDelete(e) {
+    e.stopPropagation();
```
Comment (adrianfish) [inline webcomponents/tool/src/main/frontend/packages/sakai-rubrics/src/SakaiRubricsSharedList.js:111]:
```suggestion

    e.stopPropagation();
```

File: webcomponents/tool/src/main/frontend/packages/sakai-rubrics/src/SakaiRubricReadonly.js [hunk 15]
```diff
@@ -83,4 +95,14 @@ export class SakaiRubricReadonly extends SakaiRubric {
     e.stopPropagation();
     this.dispatchEvent(new CustomEvent("copy-to-site", { detail: this.rubric.id }));
   }
+
+  deleteRubric(e) {
+    e.stopPropagation();
```
Comment (adrianfish) [inline webcomponents/tool/src/main/frontend/packages/sakai-rubrics/src/SakaiRubricReadonly.js:100]:
```suggestion

    e.stopPropagation();
```

**Review by adrianfish — APPROVED** [review]
Thanks for the work - this is looking good. I have a few small commits. I like to keep to the convention of leaving a blank line on multiline function definitions. I know it may seem a bit odd, but I think it makes js a bit more readable, especially in lambdas.
